File: java/src/main/java/lab/gabon/service/VideoService.java
package lab.gabon.service;

import java.time.Instant;
import java.util.List;
import lab.gabon.common.ApiResponse;
import lab.gabon.common.AppError;
import lab.gabon.common.AppException;
import lab.gabon.common.PageResponse;
import lab.gabon.model.entity.Video;
import lab.gabon.model.entity.VideoPlayRecord;
import lab.gabon.model.request.VideoRequests.ConfirmUploadRequest;
import lab.gabon.model.request.VideoRequests.UploadUrlRequest;
import lab.gabon.model.response.VideoResponses.UploadUrlResponse;
import lab.gabon.model.response.VideoResponses.VideoDetailResponse;
import lab.gabon.model.response.VideoResponses.VideoListItemResponse;
import lab.gabon.repository.CustomerRepository;
import lab.gabon.repository.VideoLikeRepository;
import lab.gabon.repository.VideoPlayRecordRepository;
import lab.gabon.repository.VideoRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

@Service
public class VideoService {

  private static final Logger log = LoggerFactory.getLogger(VideoService.class);

  private static final short STATUS_PENDING = 1;
  private static final short STATUS_APPROVED = 4;
  private static final short PLAY_TYPE_CLICK = 1;
  private static final short PLAY_TYPE_VALID = 2;

  private final VideoRepository videoRepo;
  private final VideoLikeRepository videoLikeRepo;
  private final VideoPlayRecordRepository playRecordRepo;
  private final CustomerRepository customerRepo;
  private final StorageService storageService;

  public VideoService(
      VideoRepository videoRepo,
      VideoLikeRepository videoLikeRepo,
      VideoPlayRecordRepository playRecordRepo,
      CustomerRepository customerRepo,
      StorageService storageService) {
    this.videoRepo = videoRepo;
    this.videoLikeRepo = videoLikeRepo;
    this.playRecordRepo = playRecordRepo;
    this.customerRepo = customerRepo;
    this.storageService = storageService;
  }

  // -- List ---------------------------------------------------------------

  public ApiResponse<PageResponse<VideoListItemResponse>> listVideos(int page, int pageSize) {
    int offset = (page - 1) * pageSize;
    var videos = videoRepo.findApprovedVideos(pageSize, offset);
    long total = videoRepo.countApprovedVideos();
    return ApiResponse.ok(new PageResponse<>(toListItems(videos), total, page, pageSize));
  }

  public ApiResponse<PageResponse<VideoListItemResponse>> listFeatured(int page, int pageSize) {
    int offset = (page - 1) * pageSize;
    var videos = videoRepo.findFeaturedVideos(pageSize, offset);
    long total = videoRepo.countApprovedVideos();
    return ApiResponse.ok(new PageResponse<>(toListItems(videos), total, page, pageSize));
  }

  public ApiResponse<PageResponse<VideoListItemResponse>> listMyVideos(
      long customerId, int page, int pageSize) {
    int offset = (page - 1) * pageSize;
    var videos = videoRepo.findMyVideos(customerId, pageSize, offset);
    long total = videoRepo.countMyVideos(customerId);
    return ApiResponse.ok(new PageResponse<>(toListItems(videos), total, page, pageSize));
  }

  public ApiResponse<PageResponse<VideoListItemResponse>> listUserVideos(
      long userId, int page, int pageSize) {
    int offset = (page - 1) * pageSize;
    var videos = videoRepo.findUserVideos(userId, pageSize, offset);
    long total = videoRepo.countUserVideos(userId);
    return ApiResponse.ok(new PageResponse<>(toListItems(videos), total, page, pageSize));
  }

  // -- Detail -------------------------------------------------------------

  public ApiResponse<VideoDetailResponse> getVideo(long videoId, Long customerId) {
    var video = videoRepo.findActiveById(videoId).orElseThrow(() -> videoNotFound());

    boolean isApproved = video.status() == STATUS_APPROVED;
    boolean isOwner = customerId != null && video.customerId() == customerId;
    if (!isApproved && !isOwner) {
      throw new AppException(new AppError.VideoNotApproved());
    }

    boolean liked =
        customerId != null && videoLikeRepo.existsByVideoIdAndCustomerId(videoId, customerId);

    var customer = customerRepo.findActiveById(video.customerId()).orElse(null);
    var detail = toDetail(video, customer, liked);
    return ApiResponse.ok(detail);
  }

  // -- Upload -------------------------------------------------------------

  public ApiResponse<UploadUrlResponse> generateUploadUrl(long customerId, UploadUrlRequest req) {
    var result = storageService.generateVideoUploadUrl(req.fileName());
    return ApiResponse.ok(result);
  }

  public ApiResponse<VideoDetailResponse> confirmUpload(long customerId, ConfirmUploadRequest req) {
    var uploadResult = storageService.generateVideoUploadUrl(req.fileName());
    var video =
        new Video(
            null,
            customerId,
            req.title(),
            req.description(),
            req.fileName(),
            req.fileSize(),
            uploadResult.fileUrl(),
            null,
            null,
            req.mimeType(),
            req.duration(),
            req.width(),
            req.height(),
            STATUS_PENDING,
            null,
            null,
            null,
            0,
            0,
            0,
            Instant.now(),
            Instant.now(),
            null);
    var saved = videoRepo.save(video);

    var customer = customerRepo.findActiveById(customerId).orElse(null);
    return ApiResponse.ok(toDetail(saved, customer, false));
  }

  // -- Delete -------------------------------------------------------------

  public ApiResponse<Void> deleteVideo(long videoId, long customerId) {
    int rows = videoRepo.softDelete(videoId, customerId);
    if (rows == 0) {
      throw videoNotFound();
    }
    return ApiResponse.ok();
  }

  // -- Like / Unlike ------------------------------------------------------

  public ApiResponse<Void> likeVideo(long videoId, long customerId) {
    int rows = videoLikeRepo.likeVideo(videoId, customerId);
    if (rows == 0) {
      throw new AppException(new AppError.AlreadyLiked());
    }
    return ApiResponse.ok();
  }

  public ApiResponse<Void> unlikeVideo(long videoId, long customerId) {
    int rows = videoLikeRepo.unlikeVideo(videoId, customerId);
    if (rows == 0) {
      log.debug("Unlike had no effect: videoId={}, customerId={}", videoId, customerId);
    }
    return ApiResponse.ok();
  }

  // -- Play Tracking ------------------------------------------------------

  public ApiResponse<Void> recordPlayClick(long videoId, Long customerId, String ipAddress) {
    videoRepo.incrementTotalClicks(videoId);
    playRecordRepo.save(
        new VideoPlayRecord(null, videoId, customerId, PLAY_TYPE_CLICK, ipAddress, Instant.now()));
    return ApiResponse.ok();
  }

  public ApiResponse<Void> recordPlayValid(long videoId, Long customerId, String ipAddress) {
    videoRepo.incrementValidClicks(videoId);
    playRecordRepo.save(
        new VideoPlayRecord(null, videoId, customerId, PLAY_TYPE_VALID, ipAddress, Instant.now()));
    return ApiResponse.ok();
  }

  // -- Internal -----------------------------------------------------------

  private List<VideoListItemResponse> toListItems(List<Video> videos) {
    return videos.stream().map(this::toListItem).toList();
  }

  private VideoListItemResponse toListItem(Video v) {
    var customer = customerRepo.findActiveById(v.customerId()).orElse(null);
    return new VideoListItemResponse(
        v.id(),
        v.customerId(),
        customer != null ? customer.name() : null,
        customer != null ? customer.avatarUrl() : null,
        v.title(),
        v.fileUrl(),
        v.thumbnailUrl(),
        v.likeCount(),
        v.totalClicks(),
        v.createdAt());
  }

  private VideoDetailResponse toDetail(
      Video v, lab.gabon.model.entity.Customer customer, boolean liked) {
    return new VideoDetailResponse(
        v.id(),
        v.customerId(),
        customer != null ? customer.name() : null,
        customer != null ? customer.avatarUrl() : null,
        v.title(),
        v.description(),
        v.fileUrl(),
        v.thumbnailUrl(),
        v.status(),
        v.totalClicks(),
        v.validClicks(),
        v.likeCount(),
        liked,
        v.createdAt());
  }

  private static AppException videoNotFound() {
    return new AppException(new AppError.VideoNotFound());
  }
}

File: kotlin/src/main/kotlin/lab/gabon/service/VideoService.kt
package lab.gabon.service

import lab.gabon.model.AppError
import lab.gabon.model.AppException
import lab.gabon.model.BUCKET_VIDEOS
import lab.gabon.model.PlayType
import lab.gabon.model.VideoStatus
import lab.gabon.repository.PlayRecordRepo
import lab.gabon.repository.VideoListRow
import lab.gabon.repository.VideoRepo
import lab.gabon.repository.VideoRow

data class PresignResult(
    val uploadUrl: String,
    val fileUrl: String,
    val s3Key: String,
)

data class VideoDetail(
    val id: Long,
    val customerId: Long,
    val title: String?,
    val description: String?,
    val fileName: String,
    val fileSize: Long,
    val fileUrl: String,
    val thumbnailUrl: String?,
    val mimeType: String,
    val duration: Int?,
    val status: Short,
    val totalClicks: Long,
    val validClicks: Long,
    val likeCount: Long,
    val createdAt: kotlinx.datetime.Instant,
    val uploaderName: String?,
    val uploaderAvatar: String?,
    val isLiked: Boolean,
)

class VideoService(
    private val videoRepo: VideoRepo,
    private val playRecordRepo: PlayRecordRepo,
    private val storageService: StorageService,
) {
    suspend fun presignUpload(
        customerId: Long,
        fileName: String,
        contentType: String,
    ): PresignResult {
        val ext = fileName.substringAfterLast('.', "bin")
        val s3Key = "$customerId/${java.util.UUID.randomUUID()}.$ext"
        val uploadUrl = storageService.presignUpload(BUCKET_VIDEOS, s3Key, contentType)
        val fileUrl = storageService.buildPublicUrl(BUCKET_VIDEOS, s3Key)
        return PresignResult(uploadUrl = uploadUrl, fileUrl = fileUrl, s3Key = s3Key)
    }

    suspend fun confirmUpload(
        customerId: Long,
        s3Key: String,
        fileName: String,
        fileSize: Long,
        mimeType: String,
        title: String?,
        description: String? = null,
    ): Long {
        val fileUrl = storageService.buildPublicUrl(BUCKET_VIDEOS, s3Key)
        return videoRepo.create(
            customerId = customerId,
            title = title,
            description = description,
            fileName = fileName,
            fileSize = fileSize,
            fileUrl = fileUrl,
            mimeType = mimeType,
            status = VideoStatus.PENDING_REVIEW.value,
        )
    }

    suspend fun listVideos(
        page: Int,
        pageSize: Int,
        keyword: String? = null,
    ): Pair<List<VideoListRow>, Long> = videoRepo.listApproved(page, pageSize, keyword)

    suspend fun listFeatured(
        page: Int,
        pageSize: Int,
    ): Pair<List<VideoListRow>, Long> = videoRepo.listFeatured(page, pageSize)

    suspend fun getDetail(
        videoId: Long,
        currentCustomerId: Long?,
    ): VideoDetail {
        val video =
            videoRepo.findById(videoId)
                ?: throw AppException(AppError.VideoNotFound())

        val isApproved = video.status == VideoStatus.APPROVED.value
        val isOwner = currentCustomerId != null && video.customerId == currentCustomerId

        if (!isApproved && !isOwner) {
            throw AppException(AppError.VideoNotApproved())
        }

        val isLiked =
            if (currentCustomerId != null) {
                videoRepo.isLikedBy(videoId, currentCustomerId)
            } else {
                false
            }

        return video.toDetail(isLiked)
    }

    suspend fun listMyVideos(
        customerId: Long,
        page: Int,
        pageSize: Int,
        status: Short? = null,
    ): Pair<List<VideoListRow>, Long> = videoRepo.listByCustomer(customerId, page, pageSize, status)

    suspend fun listUserVideos(
        userId: Long,
        page: Int,
        pageSize: Int,
    ): Pair<List<VideoListRow>, Long> = videoRepo.listApprovedByUser(userId, page, pageSize)

    suspend fun likeVideo(
        videoId: Long,
        customerId: Long,
    ) {
        val video =
            videoRepo.findById(videoId)
                ?: throw AppException(AppError.VideoNotFound())
        if (video.status != VideoStatus.APPROVED.value) {
            throw AppException(AppError.VideoNotApproved())
        }
        videoRepo.likeVideo(videoId, customerId)
    }

    suspend fun unlikeVideo(
        videoId: Long,
        customerId: Long,
    ) {
        videoRepo.unlikeVideo(videoId, customerId)
    }

    suspend fun deleteVideo(
        videoId: Long,
        customerId: Long,
    ) {
        val deleted = videoRepo.softDelete(videoId, customerId)
        if (!deleted) {
            throw AppException(AppError.VideoNotFound())
        }
    }

    suspend fun recordPlayClick(
        videoId: Long,
        customerId: Long?,
        ipAddress: String?,
    ) {
        videoRepo.incrementTotalClicks(videoId)
        playRecordRepo.create(
            videoId = videoId,
            customerId = customerId,
            playType = PlayType.CLICK.value,
            ipAddress = ipAddress,
        )
    }

    suspend fun recordValidPlay(
        videoId: Long,
        customerId: Long?,
        ipAddress: String?,
    ) {
        videoRepo.incrementValidClicks(videoId)
        playRecordRepo.create(
            videoId = videoId,
            customerId = customerId,
            playType = PlayType.VALID_PLAY.value,
            ipAddress = ipAddress,
        )
    }

    private fun VideoRow.toDetail(isLiked: Boolean): VideoDetail =
        VideoDetail(
            id = id,
            customerId = customerId,
            title = title,
            description = description,
            fileName = fileName,
            fileSize = fileSize,
            fileUrl = fileUrl,
            thumbnailUrl = thumbnailUrl,
            mimeType = mimeType,
            duration = duration,
            status = status,
            totalClicks = totalClicks,
            validClicks = validClicks,
            likeCount = likeCount,
            createdAt = createdAt,
            uploaderName = uploaderName,
            uploaderAvatar = uploaderAvatar,
            isLiked = isLiked,
        )
}
